components/fonts/font_store.rs

```diff
@@ -3,25 +3,72 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
 use std::collections::{HashMap, HashSet};
-use std::sync::Arc;
+use std::sync::{Arc, OnceLock};
 
 use app_units::Au;
 use atomic_refcell::AtomicRefCell;
+use ipc_channel::ipc::IpcSharedMemory;
 use log::warn;
 use parking_lot::RwLock;
+use serde::{Deserialize, Serialize};
 use style::stylesheets::DocumentStyleSheet;
 use style::values::computed::{FontStyle, FontWeight};
 use webrender_api::{FontInstanceFlags, FontInstanceKey, FontKey};
 
 use crate::font::FontDescriptor;
-use crate::font_cache_thread::{FontIdentifier, FontSource, LowercaseFontFamilyName};
+use crate::font_cache_thread::{
+    FontIdentifier, LowercaseFontFamilyName, SystemFontServiceProxyTrait,
+};
 use crate::font_context::WebFontDownloadState;
 use crate::font_template::{FontTemplate, FontTemplateRef, FontTemplateRefMethods, IsOblique};
+use crate::FontContext;
+
+/// A data structure to store data for fonts. If sent across IPC channels and only a
+/// [`IpcSharedMemory`] handle is sent, avoiding the overhead of serialization and
+/// deserialization. In addition, if a shared handle to data is requested
+/// (`Arc<Vec<u8>>`), the data is lazily copied out of shared memory once per
+/// [`FontData`].
+#[derive(Debug, Deserialize, Serialize)]
+pub struct FontData {
+    /// The data of this font in shared memory. Suitable for sending across IPC channels.
+    shared_memory: Arc<IpcSharedMemory>,
+    /// A lazily-initialized copy of the data behind an [`Arc`] which can be used when
+    /// passing it to various APIs.
+    #[serde(skip)]
+    arc: OnceLock<Arc<Vec<u8>>>,
```

Comment:
I would say `Arc<OnceLock<_>>` would make more sense, which would allow the arc in `Arc<FontFata>` to be removed. 

By the way I don't understand the need for `Arc<Vec<u8>>`, why not just an `Rc<>`? (unlike the shared mem, the vec is not send or sync).

Reply:
Ok, thanks for the info.